src/query/api/v1/handler/prometheus/debugger/validate.go

```diff
@@ -0,0 +1,269 @@
+	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
+	logger := logging.WithContext(ctx)
+
+	defer r.Body.Close()
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		logger.Error("unable to read data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	var promDebug prometheus.PromDebug
+	if err := json.Unmarshal(data, &promDebug); err != nil {
+		logger.Error("unable to unmarshal data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	s, err := debug.NewStorage(promDebug.Input)
+	if err != nil {
+		logger.Error("unable to create storage", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	h.engine = executor.NewEngine(s, h.scope.SubScope("debug_engine"))
+	results, _, err := h.readHandler.ServeHTTPWithEngine(w, r, h.engine)
+	if err != nil {
+		logger.Error("unable to read data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	promResults, err := debug.PromResultToSeriesList(promDebug.Results, models.NewTagOptions())
+	if err != nil {
+		logger.Error("unable to convert prom results data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	mismatches, err := validate(tsListToMap(promResults), tsListToMap(results))
+	if err != nil {
+		logger.Error("error validating results", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+
+	if err := renderDebugMismatchResultsJSON(w, mismatches); err != nil {
+		logger.Error("unable to write back mismatch data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+}
+
+func tsListToMap(tsList []*ts.Series) map[string]*ts.Series {
+	tsMap := make(map[string]*ts.Series, len(tsList))
+
+	for _, series := range tsList {
+		series.Tags = series.Tags.Normalize()
+		id := series.Tags.ID()
+		tsMap[id] = series
+	}
+
+	return tsMap
+}
+
+type mismatch struct {
+	seriesName       string
+	promVal, m3Val   float64
+	promTime, m3Time time.Time
+}
+
+func validate(prom, m3 map[string]*ts.Series) ([][]mismatch, error) {
+	if len(prom) != len(m3) {
+		return nil, errors.New("number of Prom series not equal to number of M3 series")
+	}
+
+	mismatches := make([][]mismatch, 0, mismatchCapacity)
+
+	for id, promSeries := range prom {
+		var (
+			m3Series *ts.Series
+			exists   bool
+		)
+
+		if m3Series, exists = m3[id]; !exists {
+			return nil, fmt.Errorf("series with id %s does not exist in M3 results", id)
+		}
+
+		promdps := promSeries.Values().Datapoints()
+		m3dps := m3Series.Values().Datapoints()
+
+		mismatchList := make([]mismatch, 0, mismatchCapacity)
+
+		m3idx := 0
+		m3dp := m3dps[m3idx]
+
+		for _, promdp := range promdps {
+			if math.IsNaN(promdp.Value) && !math.IsNaN(m3dp.Value) {
+				mismatchList = append(mismatchList, createMismatch(id, promdp.Value, m3dp.Value, promdp.Timestamp, m3dp.Timestamp))
+			}
+
+			for {
```

Comment:
Nit: I think this would be equivalently readable as a `for` condition, maybe with a comment

```
// advance past NaN's
for ; m3idx < len(m3dps) && math.IsNaN(m3dps[m3idx].Value); m3idx++ {}
```

Reply:
I like it